### PR TITLE
Utility functions for extended bv rewriter

### PR DESCRIPTION
**Pull request by ajreynol**
First step towards merging https://github.com/ajreynol/CVC4/tree/extRewBv.

### CODE REVIEW

File: src/theory/quantifiers/extended_rewrite.cpp [hunk 1]
```diff
@@ -1467,9 +1468,30 @@ Node ExtendedRewriter::partialSubstitute(Node n,
...
+    {
+      if (isConstBv(n[1 - i], false))
+      {
+        // (bvadd x (bvneg y)) = 0 ---> x = y
+        if (n[i].getKind() == BITVECTOR_PLUS && n[i].getNumChildren() == 2)
+        {
+          for (unsigned j = 0; j < 2; j++)
+          {
+            if (n[i][j].isVar() && n[i][1 - j].getKind() == BITVECTOR_NEG
+                && n[i][1 - j][0].isVar())
```
Comment (aniemetz) [inline src/theory/quantifiers/extended_rewrite.cpp]: Why is this limited to vars only?

File: src/theory/quantifiers/extended_rewrite.cpp [hunk 2]
```diff
@@ -1563,6 +1585,571 @@ Node ExtendedRewriter::extendedRewriteArith(Node ret)
...
 
+int ExtendedRewriter::bitVectorSubsume(Node a, Node b, bool strict, bool tryNot)
+{
+  Assert(a.getType() == b.getType());
+  int max_ret = strict ? 2 : 1;
+  int curr_ret = 0;
+  Trace("q-ext-rewrite-debug2")
+      << "Subsume " << a << " " << b << "?" << std::endl;
+  // simple cases : a=b, a=max, b=0
+  if (a == b || isConstBv(a, true) || isConstBv(b, false))
```
Comment (aniemetz) [inline src/theory/quantifiers/extended_rewrite.cpp]: `isConstBv` only tests for zero and ones (depending on flag `isNot`), IMO its name is quite misleading.
Also, it is actually not really necessary, you could use `bv::utils::isZero()` to test for zero (https://github.com/CVC4/CVC4/blob/master/src/theory/bv/theory_bv_utils.h#L56) and add a corresponding `isOnes()`. That would be way cleaner and `isConstBv` would then be obsolete.

File: src/theory/quantifiers/extended_rewrite.cpp [hunk 3]
```diff
@@ -1563,6 +1585,571 @@ Node ExtendedRewriter::extendedRewriteArith(Node ret)
...
+  return curr;
+}
+
+Node ExtendedRewriter::mkConstBv(Node n, bool isNot)
+{
+  unsigned size = bv::utils::getSize(n);
+  return isNot ? bv::utils::mkOnes(size) : bv::utils::mkZero(size);
+}
+
+bool ExtendedRewriter::isConstBv(Node n, bool isNot)
```
Comment (aniemetz) [inline src/theory/quantifiers/extended_rewrite.cpp]: See above.

File: src/theory/quantifiers/extended_rewrite.cpp [hunk 4]
```diff
@@ -1563,6 +1585,571 @@ Node ExtendedRewriter::extendedRewriteArith(Node ret)
...
+    }
+    else
+    {
+      nconst.push_back(cn);
+    }
+  }
+  return ret;
+}
+
+bool ExtendedRewriter::hasConstBvChild(Node n)
```
Comment (aniemetz) [inline src/theory/quantifiers/extended_rewrite.cpp]: I'm wondering if this should actually be a `bv::utils` function.

File: src/theory/quantifiers/extended_rewrite.cpp [hunk 5]
```diff
@@ -1563,6 +1585,571 @@ Node ExtendedRewriter::extendedRewriteArith(Node ret)
...
+            << "...more than one bit diff." << std::endl;
+        return -1;
+      }
+      bit_diff_index = i;
+    }
+  }
+  if (bit_diff_index >= 0)
+  {
+    std::vector<Node> split;
+    if (bit_diff_index + 1 < static_cast<int>(csize))
```
Comment (aniemetz) [inline src/theory/quantifiers/extended_rewrite.cpp]: This is way ugly. I'd suggest to use `unsigned` for `bit_diff_index` and use a flag that indicates if it has been set.

File: src/theory/quantifiers/extended_rewrite.cpp [hunk 6]
```diff
@@ -1563,6 +1585,571 @@ Node ExtendedRewriter::extendedRewriteArith(Node ret)
...
+    diff_index = nv.size();
+    nv.push_back(bit);
+    // remainder
+    if (bit_diff_index > 0)
+    {
+      Node extract = bv::utils::mkExtract(c1, bit_diff_index - 1, 0);
+      nv.push_back(Rewriter::rewrite(extract));
+    }
+    // insert suffix
+    if (diff_index < static_cast<int>(rc1.size()))
```
Comment (aniemetz) [inline src/theory/quantifiers/extended_rewrite.cpp]: Ditto (for `diff_index`)

File: src/theory/quantifiers/extended_rewrite.cpp [hunk 7]
```diff
@@ -1563,6 +1585,571 @@ Node ExtendedRewriter::extendedRewriteArith(Node ret)
...
+  Trace("q-ext-rewrite-debug") << "extendBv " << n << std::endl;
+  std::vector<Node> children;
+  int counter = bv::utils::getSize(n) - 1;
+  for (const std::pair<const unsigned, Node>& ep : ex_map)
+  {
+    Trace("q-ext-rewrite-debug")
+        << "  process " << ep.first << " : " << ep.second
+        << ", counter=" << counter << std::endl;
+    unsigned start = ep.first;
+    Assert(static_cast<int>(start) <= counter);
```
Comment (aniemetz) [inline src/theory/quantifiers/extended_rewrite.cpp]: Ditto (for `counter`)

File: src/theory/quantifiers/extended_rewrite.h [hunk 8]
```diff
@@ -228,6 +229,107 @@ class ExtendedRewriter
...
+   *   n1 is equivalent to nv[0] ++ ... ++ nv[i] ++ ... ++ nv[k-1],
+   *   n2 is equivalent to nv[0] ++ ... ++ (~)nv[i] ++ ... ++ nv[k-1], and
+   *   nv[i] is a constant of bit-width one.
+   */
+  int spliceBvConstBit(Node n1, Node n2, std::vector<Node>& nv);
+  /** extend bit-vector
+   *
+   * This returns the concatentation node of the form
+   *   concat( ((_ extract s1 e1) n) ... ((_ extract sn en) n))
+   * where s1 = bitwidth(n)-1, s_{i+1} = e_i - 1 for each i=2,...n,
```
Comment (aniemetz) [inline src/theory/quantifiers/extended_rewrite.h]: e_i  = ?

**Comment by aniemetz** [comment]
Also, I'd like to request a review from @mpreiner here.

File: src/theory/quantifiers/extended_rewrite.cpp [hunk 9]
```diff
@@ -1563,6 +1585,571 @@ Node ExtendedRewriter::extendedRewriteArith(Node ret)
...
+  if (!rec.isNull())
+  {
+    int r1 = bitVectorSubsume(reca ? a[1] : a, reca ? b : b[1], strict);
+    if (r1 > curr_ret)
+    {
+      int r2 = bitVectorSubsume(reca ? a[2] : a, reca ? b : b[2], strict);
+      if (r2 > curr_ret)
+      {
+        curr_ret = r2 > r1 ? r1 : r2;
+        if (curr_ret = max_ret)
```
Comment (4tXJ7f) [inline src/theory/quantifiers/extended_rewrite.cpp]: This should be `curr_ret == max_reg`, no?

Comment (ajreynol) [reply src/theory/quantifiers/extended_rewrite.cpp]: Yes, ill fix

**Comment by ajreynol** [comment]
Closing this for now, this PR needs revisiting and was lower priority.

**Pull request closed without merging**